### DISCUSSION TITLE
Symbolic Transformers

### DISCUSSION THREAD

**Discussion by LukasZahradnik** [LukasZahradnik/PyNeuraLogic#37] [Announcements]
## Additions:

- Transformer modules - `Transformer`, `TransformerEncoder`, and `TransformerDecoder`.
- Attention modules - `Attention`, `MultiheadAttention`.
- `PositionalEncoding` module.
- Softmax as a (splittable) aggregation function.
- Concatenation as an aggregation function.
- Reshape transformation function.
- Slice transformation function.
- Learning rate decay as an optimizer argument.
- Batching support.
- Progress bar for tracking dataset-building progress.
- Special arithmetic predicates for addition (`add`), subtraction (`sub`), and modulo (`mod`).

## Changes:

- Concatenation can now be parametrized with a dimension.
- Refactored MLP module - the current version is more compact.
- Changed the behavior of `Variable`/`V` factory - only the first letter is capitalized.
- Performance improvements.

<hr /><em>This discussion was created from the release <a href='https://github.com/LukasZahradnik/PyNeuraLogic/releases/tag/v0.7.0'>Symbolic Transformers</a>.</em>

**Comment by GustikS** [comment]
this will be huge, we really need to find ways to propagate it...

**Comment by joaquincabezas** [comment]
I was just checking that nothing had changed from October and then I found this! Amazing :)

do you have any result/benchmark on this new feature?

**Comment by LukasZahradnik** [comment]
@joaquincabezas We only have [benchmarks](https://pyneuralogic.readthedocs.io/en/latest/benchmarks.html) for GNN models, although those need to be updated, as we made a lot of performance improvements. But I'm considering adding benchmarks on transformers.

**Reply by joaquincabezas** [reply]
At this point, I think maybe practical example on using Transformers within the framework would be more useful than the benchmark itself. I could elaborate some examples as well.

**Reply by LukasZahradnik** [reply]
@joaquincabezas Agree. We plan to focus more on articles/examples now, and having some help with that from you would be awesome.

I'm currently in the process of writing an article about Transformers, but that is more about the motivation to use symbolic representation and about implementing Transformers from scratch (rather than some of its concrete usage).

**Reply by joaquincabezas** [reply]
what concrete usage do you think may worth the effort? I could focus on that (maybe some paper or an example from another framework that can be replicated...)

**Reply by GustikS** [reply]
I think we could eventually do something like https://arxiv.org/abs/2301.05062 - i.e. providing a "language" for mapping between Transformers and classic algorithms - that seems to be gaining a lot of hype now, and we can actually do more than that from expressiveness and interpretability points of view (as we also have recursion, logical constraints, etc.).

**Reply by joaquincabezas** [reply]
I was not aware of this trend! In fact the approach proposed in that paper is completely novel for me. I am still reading it but I think this can work for my PhD also, so instead of a light contribution I can do some heavy work.

**Reply by GustikS** [reply]
Well this was just put on arxiv a few days ago, not sure it will survive into a longer-term trend, but it is a line of thinking closely related to this framework, the way I see it. But we surely need to start with some concrete examples and use-cases first while improving the UX, I guess it will be a bit clearer once Lukas comes with the tutorial...

**Reply by LukasZahradnik** [reply]
@joaquincabezas I was thinking about replicating [the tutorial that PyTorch has](https://pytorch.org/tutorials/beginner/transformer_tutorial.html). But that might be too large a dataset for PyNeuraLogic (I'm not sure), so maybe something smaller, like a translation of roman numerals, etc., would be sufficient.

**Reply by joaquincabezas** [reply]
You're right. But these days to publish something on ML you either need to go into the unknown or assume the risk of becoming obsolete before publishing.

@LukasZahradnik how can I help you with the tutorial?

**Reply by joaquincabezas** [reply]
Oh my previous answer was for the message https://github.com/LukasZahradnik/PyNeuraLogic/discussions/37#discussioncomment-4690669 but I clicked the Reply button after a while. So you already answered my question in the past!

I like the roman numbers idea, I saw a sequence2sequence example in Matlab and it also is a clear one for a tutorial
